### PR TITLE
tracing: instrument more resources

### PR DESCRIPTION
**Pull request by zaharidichev**
This PR adds instrumentation to more resources from the `sync` package.
Related: https://github.com/tokio-rs/console/pull/188

### CODE REVIEW

File: tokio/src/sync/batch_semaphore.rs [hunk 1]
```diff
@@ -150,12 +185,25 @@ impl Semaphore {
...
+
+        cfg_not_trace! {
+            return Self {
+                permits: AtomicUsize::new(permits << Self::PERMIT_SHIFT),
+                waiters: Mutex::const_new(Waitlist {
+                    queue: LinkedList::new(),
+                    closed: false,
+                }),
+            };
         }
```
Comment (Darksonn) [inline tokio/src/sync/batch_semaphore.rs:207]: Surely there must be a better way...

Comment (zaharidichev) [reply tokio/src/sync/batch_semaphore.rs:207]: I would like to believe so. Any ideas?

Comment (Darksonn) [reply tokio/src/sync/batch_semaphore.rs:207]: What happens if you do this?
```Rust
return Self {
    permits: AtomicUsize::new(permits << Self::PERMIT_SHIFT),
    waiters: Mutex::const_new(Waitlist {
        queue: LinkedList::new(),
        closed: false,
    }),
    #[cfg(tracing)]
    resource_span: tracing::Span::none(),
};
```

Comment (zaharidichev) [reply tokio/src/sync/batch_semaphore.rs]: Thanks a lot, I have changed it to use this approach in other places as well

**Review by carllerche — APPROVED** [review]
Thanks for this. Looks good to merge.

As a general comment, there probably are opportunities to clean up the code and avoid all the `cfg` attributes in the main source. We can worry about that in a follow up PR. I'm happy to work with you on figuring out strategies to clean it up.

The change here will only apply when the unstable flag is enabled, so lets move forward and cleanup incrementally.

**Pull request merged**